src/storage-client/src/source/persist_source.rs

```diff
@@ -528,18 +588,270 @@ impl PersistSourceDataStats<'_> {
+                        } else {
+                            i += 1;
+                        }
+                    }
+                    yield Either::Left(prog)
+                }
+            }
+        }
+    });
+    let shutdown_button = builder.build(move |caps| async move {
+        // The output capability.
+        let mut cap_set = CapabilitySet::from_elem(caps.into_element());
+
+        // The frontier of our output. This matches the `CapabilitySet` above.
+        let mut output_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+        // The frontier of the `flow_control` input.
+        let mut flow_control_frontier = Antichain::from_elem(TimelyTimestamp::minimum());
+
+        // Parts we have emitted, but have not yet retired (based on the `flow_control` edge).
+        let mut inflight_parts = Vec::new();
+        // Parts we have not yet emitted, but do participate in the `input_frontier`.
+        let mut pending_parts = std::collections::VecDeque::new();
+
+        // Only one worker is responsible for distributing parts
+        if worker_index != chosen_worker {
+            trace!(
+                "We are not the chosen worker ({}), exiting...",
+                chosen_worker
+            );
+            return;
+        }
+        tokio::pin!(data_input);
+        'emitting_parts: loop {
+            // At the beginning of our main loop, we determine the total size of
+            // inflight parts.
+            let inflight_bytes: usize = inflight_parts.iter().map(|(_, size)| size).sum();
+
+            // There are 2 main cases where we can continue to emit parts:
+            // - The total emitted bytes is less than `flow_control_max_bytes`.
+            // - We have not yet reached the `flow_control_frontier`
+            //
+            // SUBTLE: in the latter case, we may arbitrarily go into the backpressure `else`
+            // block, as we wait for progress tracking to keep the `flow_control` frontier
+            // up-to-date. This is tested in unit-tests.
+            if inflight_bytes < flow_control_max_bytes
+                || PartialOrder::less_than(&output_frontier, &flow_control_frontier)
+            {
+                let (time, part) = if let Some((time, part)) = pending_parts.pop_front() {
+                    (time, part)
+                } else {
+                    match data_input.next().await {
+                        Some(Either::Right((time, part))) => {
+                            // Downgrade the output frontier to this part's time. This is useful
+                            // "close" timestamp's from previous parts, even if we don't yet
+                            // emit this part. Note that this is safe because `data_input` ensures
+                            // time-ordering.
+                            output_frontier = Antichain::from_elem(time.clone());
+                            cap_set.downgrade(output_frontier.iter());
```

Comment:
This is not safe under `PartialOrder` and even though we have the `TotalOrder` bound we want to remove it so let's not add more ties to it. I would suggest that we only rely on progress statements. In order to do this correctly you need to wait for the frontier of the descriptions to advance, which would let you form a batch of descriptions, and then you can play back the descriptions and synthesize fine grained frontier updates.

Reply:
the `data_input` stream maintains this order, so in some sense I AM synthesizing fine grained updates. Will update to use `Antichain::insert` without actual progress frontiers, let me know what you think. If this gets that much more complicated, we should merge with the `TotalOrder` and followup with a `PartialOrder`-ification (I know you hate this but we are on a time crunch on this)